NDTensors/src/lib/Sectors/src/category_product.jl

```diff
@@ -34,78 +52,224 @@ function Base.show(io::IO, s::CategoryProduct)
+  v = map(
+    ((l1, l2),) -> l1 × l2,
+    Iterators.flatten((Iterators.product(blocklengths(g1), blocklengths(g2)),),),
+  )
+  return gradedrange(v)
+end
+
+# ====================================  Fusion rules  ======================================
+# generic case: fusion returns a GradedAxes, even for fusion with Empty
+function fusion_rule(::SymmetryStyle, s1::CategoryProduct, s2::CategoryProduct)
+  return to_graded_axis(categories_fusion_rule(categories(s1), categories(s2)))
+end
+
+# Abelian case: fusion returns CategoryProduct
+function fusion_rule(::AbelianGroup, s1::CategoryProduct, s2::CategoryProduct)
+  return categories_fusion_rule(categories(s1), categories(s2))
+end
+
+# Empty case
+function fusion_rule(::EmptyCategory, ::EmptyCategoryProduct, ::EmptyCategoryProduct)
+  return sector()
+end
+
+# EmptyCategory acts as trivial on any AbstractCategory, not just CategoryProduct
+function fusion_rule(::SymmetryStyle, ::EmptyCategoryProduct, c::AbstractCategory)
+  return to_graded_axis(c)
+end
+function fusion_rule(::SymmetryStyle, ::EmptyCategoryProduct, c::CategoryProduct)
+  return to_graded_axis(c)
+end
+function fusion_rule(::SymmetryStyle, c::AbstractCategory, ::EmptyCategoryProduct)
+  return to_graded_axis(c)
+end
+function fusion_rule(::SymmetryStyle, c::CategoryProduct, ::EmptyCategoryProduct)
+  return to_graded_axis(c)
+end
 
+# abelian case: return Category
+fusion_rule(::AbelianGroup, ::EmptyCategoryProduct, c::AbstractCategory) = c
+fusion_rule(::AbelianGroup, ::EmptyCategoryProduct, c::CategoryProduct) = c
+fusion_rule(::AbelianGroup, c::AbstractCategory, ::EmptyCategoryProduct) = c
+fusion_rule(::AbelianGroup, c::CategoryProduct, ::EmptyCategoryProduct) = c
+
+# ===============================  Ordered implementation  =================================
+CategoryProduct(t::Tuple) = _CategoryProduct(t)
+CategoryProduct(cats::AbstractCategory...) = CategoryProduct((cats...,))
+
+categories_isequal(o1::Tuple, o2::Tuple) = (o1 == o2)
+
+categories_isless(::Tuple, ::Tuple) = throw(ArgumentError("Not implemented"))
+categories_isless(t1::T, t2::T) where {T<:Tuple} = t1 < t2
+
+categories_product(l1::Tuple, l2::Tuple) = (l1..., l2...)
+
+categories_trivial(type::Type{<:Tuple}) = trivial.(fieldtypes(type))
+
+function categories_common(t1::Tuple, t2::Tuple)
+  n = min(length(t1), length(t2))
+  return t1[begin:n], t2[begin:n]
+end
+
+function categories_diff(t1::Tuple, t2::Tuple)
+  n1 = length(t1)
+  n2 = length(t2)
+  return n1 < n2 ? t2[(n1 + 1):end] : t1[(n2 + 1):end]
+end
+
+# ===========================  Dictionary-like implementation  =============================
 function CategoryProduct(nt::NamedTuple)
   categories = sort_keys(nt)
   return _CategoryProduct(categories)
 end
 
 CategoryProduct(; kws...) = CategoryProduct((; kws...))
 
+# avoid having 2 different kinds of EmptyCategory: cast empty NamedTuple to Tuple{}
+CategoryProduct(::NamedTuple{()}) = CategoryProduct(())
```

Comment:
I'm not so sure I like this type of conversion happening automatically, what if we want `CategoryProduct((;))` for whatever reason?

If we define `EmptyCategoryProduct` like I suggest above (https://github.com/ITensor/ITensors.jl/pull/1363/files#r1779717281), can we handle both empty cases without too much trouble?

Reply:
@emstoudenmire curious to hear your thoughts on this.